File: Week 4/car_plates_dictionary.py
def find_shortest_word(plate, vocabulary):
    """Return the shortest word in the vocabulary with all the letters in a license plate."""

    # Filter the 'plate' string, grabing only letters in lowercase (a-z)
    word_list = []
    plate = plate.lower()
    for i in plate:
        if (ord(i) in range(97, 123)):
            word_list.append(i)

    # Iterate through the list of vocabulary and compare the letters in the plate with each vocabulary
    matching_vocabs = []
    if vocabulary == []:
        return ""

    for vocab in vocabulary:
        vocab = vocab.lower()
        matching_counter = 0
        seen = []
        for char in word_list:
            if char in vocab:
                if char in seen and vocab.count(char) == word_list.count(char):
                    matching_counter += 2
                else:
                    matching_counter += 1
                    seen.append(char)
            else:
                matching_counter = 0
                break
        matching_vocabs.append(tuple((matching_counter, vocab)))

    # Grab the highest number of matched letters
    max_counter = max([x[0] for x in matching_vocabs])

    if max_counter == 0:
        return ""

    # Filter the vocabulary, grabing only the ones with the highest max_counter
    final_list = []
    for pair in matching_vocabs:
        if (pair[0] == max_counter):
            final_list.append(pair[1])

    # Return the shortest word matched with the letters in the 'plate' string
    return min(final_list, key=len)

File: Week 4/carPlatesDictionary.java
// NOT YET FINISHED, TESTS ARE STILL FAILING

import java.util.List;
import java.util.ArrayList;
import org.springframework.util.StringUtils;
import java.util.Collections;

public class ShortestWordInCarPlate {
    private List<Character> wordList = new ArrayList<Character>();
    private Pair<Integer, String> matchingVocabs;
    private List<Character> seenChar;
    private int matchingCounter = 0;
    private List<String> finalList = new ArrayList<String>();

    public static String findShortestWord(String carPlate, List<String> vocabulary) {
        // Filter the 'plate' string, grabing only letters in lowercase (a-z)
        carPlate = carPlate.toLowerCase();
        for (int i=0; i < carPlate.length(); i++) {
            if ((int) carPlate.charAt(i) > 96 && (int) carPlate.charAt(i) < 123) {
                wordList.add(carPlate.charAt(i));
            }
        }

        // Iterate through the list of vocabulary and compare the letters in the plate with each vocabulary
        if (vocabulary.size() == 0) {
            return "";
        }

        for (int i=0; i < vocabulary.size(); i++) {
            vocabulary.get(i) = vocabulary.get(i).toLowerCase();
            matchingCounter = 0;
            seenChar = new ArrayList<Character>();
            for (int j=0; j < wordList.size(); j++) {
                // Count the number of occurences of a character in each vocabulary and in wordList
                int countInVocabulary = StringUtils.countOccurrencesOf(vocabulary[i], wordList[j]);
                int countInwordList = Collections.frequency(wordList, wordList[j]); 
                if (vocabulary.get(i).contains(wordList.get(j))) {
                    if (seenChar.contains(wordList.get(j)) && countInVocabulary == countInwordList) {
                        matchingCounter += 2;
                    }
                    else {
                        matchingCounter++;
                    }
                }
                else {
                    matchingCounter = 0;
                    break;
                }
            matchingVocabs.add(Pair.with(matchingCounter, vocabulary.get(i)));
            }
        }

        // Grab the highest number of matched letters
        int maxCounter = 0;
        for (int counter=0; counter < matchingVocabs.size(); counter++) {
            if (matchingVocabs.getValue(counter).getValue0() > maxCounter)
                maxCounter = matchingVocabs.getValue(counter).getValue0();
        }

        if (maxCounter == 0) {
            return "";
        }

        // Filter the vocabulary, grabing only the ones with the highest max_counter
        for (int pair=0; pair < matchingVocabs.size(); pair++) {
            if (matchingVocabs.getValue(pair).getValue0() == maxCounter)
                finalList.add(matchingVocabs.getValue(pair).getValue1());
        }

        // Return the shortest word matched with the letters in the 'plate' string
        String shortest = finalList.get(0);
        for (String element : finalList) {
            if (element.length() < shortest.length()) {
                shortest = element;
            }
        }
        return shortest;
    }
}